### PR TITLE
General Python cleanup and proof-reading

### PR DESCRIPTION
**Pull request by dberzano**
@chrisburr this is still a WIP :-)

### CODE REVIEW

**Comment by dberzano** [comment]
@chrisburr @alexpearce note that the dictionaries one contains a potentially controversial note on the usage of `items()`. I don't know what's the LHCb convention on this, but `items()` is inefficient on Python 2 (the version you use, I understand), and if looping over many many items you can feel the impact on both memory usage and CPU. This is why I have added a note essentially refraining users from making use of `items()` in Python 2. Any comment on that?

**Comment by alexpearce** [comment]
I used to be pro-`iteritems` but now rarely use it, in the spirit of writing code that's more easily portable to Python 3<sup>1</sup>. Using `items` with Python 2 is nice because it just gives you a list<sup>2</sup>, whereas with `iteritems` you'd then need to explain generators (or say something like 'this is technical detail', but I really don't look doing that). Plus, I've never had a dictionary big enough that the extra memory usage is a problem, so my use of `iteritems` was only ever a premature optimisation.

<sup>1</sup>: I did also consider also recommending always using `from __future__ import print_function`, for a similar reason, but the `__future__` thing is a bit intimidating.
<sup>2</sup>: Ironically, Python 3 would actually be harder to teach because it uses generators so extensively, I think.

File: python/dictionaries.md [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
 items.
 
-Dictionaries are like lists, but they can map from more than just integers.  
-Because of this, you must always specify the index for every item when creating 
-a dictionary.
+Dictionaries are collections, just like lists, but they have important
+differences:
+
+* lists map sequential numeric indices to items, whereas dictionaries map
+  anything to anything,
```
Comment (alexpearce) [inline python/dictionaries.md]: Technically not "anything", but the detail behind why is a bit gross 🤔

File: python/dictionaries.md [hunk 2]
```diff
@@ -68,14 +77,17 @@ We can create dictionaries from lists of 2-item lists.
...
-And also with _dictionary comprehensions_, in a similar manner to list 
+And also with _dictionary comprehensions_, in a similar manner to list
 comprehensions, with the additional specification of the key.
 
 ```python
 >>> {i: i**i for i in range(5) if i != 3}
 {0: 1, 1: 1, 2: 4, 4: 256, 5: 3125}
 ```
 
+Note that dictionary comprehensions require at least Python 2.7 to work.
```
Comment (alexpearce) [inline python/dictionaries.md:88]: Do we think compatibility with 2.6 is important?

**Review by alexpearce — COMMENTED** [review]
Thanks a lot for going over things. The lessons were written in quite a rush 😅

File: python/dictionaries.md (hunk 1 above)
Comment (dberzano) [reply python/dictionaries.md]: Yes the detail actually follows...

File: python/dictionaries.md (hunk 2 above)
Comment (dberzano) [reply python/dictionaries.md:88]: It depends. I am just thinking about the poor man connecting to lxplus, trying dictionary comprehensions and seeing that they don't work for some reason because she's forgotten to load the correct environment. I don't think this note makes much harm :-)

**Comment by chrisburr** [comment]
@dberzano I rarely see cases where it actually matters from an efficiency perspective so I personally prefer to use `items()` so that it is compatible with both Python 2 and 3. That said, I use Python 3 as much as possible so I might not see the problems as frequently. 

The solution you suggest partially suffers from the same problem as `keys()` as it also creates a list containing copies[1]. The "correct" way to write it efficiently for Python 2.7 is to use `viewitems()` which returns a generator of references instead.


[1] `viewkeys()` and `viewitems()` can be used to give the Python 3-like behaviour, but this is only compatible with Python 2.7.

File: python/dictionaries.md (hunk 1 above)
Comment (alexpearce) [reply python/dictionaries.md]: Indeed, so this creates a contradiction. Maybe "almost anything"?

Comment (dberzano) [reply python/dictionaries.md]: Let me rephrase ;-)

File: python/dictionaries.md (hunk 2 above)
Comment (alexpearce) [reply python/dictionaries.md:88]: Yeah, it's really annoying the default for `ssh lxplus` users is still 2.6... OK, let's keep it.

**Comment by dberzano** [comment]
@chrisburr @alexpearce I think it is useful to point out that instructions can have efficiency issues that can be easily overcome. It's a callout so it can be easily skipped, and for the tutorial purposes and most everyday tasks, I think that `items()` is perfectly acceptable.

But if some user at some point has her jobs killed on the Grid as they use too much memory, she might get back to those pages and find out that she's got an easy alternative to reduce memory usage. Say, I think the callout makes a good "future reference" :-)

Would you keep it like this, remove, or rephrase?

**Comment by dberzano** [comment]
By the way thanks for considering "pineapple pizza" a "weird" pizza :+1:

**Comment by alexpearce** [comment]
I think we could rephrase. How about something like this.

> In Python 2, using `items` copies the keys and values of a dictionary, and gives you back those copies. This can be problematic for large dictionaries as the amount of memory your program uses can double. Python 3 uses a much more memory-efficient way of implementing `items` so that you don't have to worry.
>
> If you're having memory problems with using `items` in Python 2, you can use `viewitems` instead, which behaves the same way as `items` does in Python 3.
>
> Note that there are also similar methods for `keys` and `values`, called `viewkeys` and `viewvalues`, and that all of these `view` methods are only available from Python 2.7.

**Comment by dberzano** [comment]
OK given that's past 8pm I'd say it's fine for now. I'll be adding more stuff after dinner, but I think this PR is complete for now and - if you agree - it can be merged. Removing the WIP...

**p.s.** I am really closing both eyes for the "garlic on pizza" example :-)

File: python/conditions.md [hunk 3]
```diff
@@ -20,26 +19,30 @@ Amazing pizza!
...
 ```python
 >>> x = 'ok' if pizzas[0] == 'Cheese' else 'not ok'
 >>> x
 'not ok'
 ```
 
-Python evaluates a condition and sees whether it truth-like or not. If it is 
+Make sure your line does not get too long in order not to impair its
+readability!
+{% endcallout %}
```
Comment (alexpearce) [inline python/conditions.md:42]: Making this a callout is a good idea 👍

**Comment by alexpearce** [comment]
Really great stuff, thanks a lot.

Please replace garlic with something tastier, I don't know what I was thinking! 😷
